pkg/generators/rules/lifecycle_tag.go

```diff
@@ -0,0 +1,242 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rules
+
+import (
+	"go/ast"
+	"go/parser"
+	"go/token"
+	gotypes "go/types"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"k8s.io/gengo/examples/set-gen/sets"
+	"k8s.io/gengo/types"
+)
+
+const (
+	// TagLifecycle is the comment tag prefix for specifying
+	// the API lifecycle info. Example tag format:
+	// +lifecycle:component=<component-name>,minVersion=<version>,status=<prerelease-status>,featureGate=<featuregate-name>
+	TagLifecycle = "lifecycle:component"
+
+	TagComponent   = "component"
+	TagMinVersion  = "minVersion"
+	TagStatus      = "status"
+	TagFeatureGate = "featureGate"
+
+	kubernetesComponentName = "kubernetes"
+
+	// as defined in https://github.com/kubernetes/kubernetes/blob/master/pkg/features/kube_features.go
+	kubernetesFeatureGateType = "featuregate.Feature"
+)
+
+var (
+	AllowedTagKeys = sets.NewString(TagComponent, TagMinVersion, TagStatus, TagFeatureGate)
+
+	// the status names must match the prereleases defined in
+	// k8s.io/component-base/featuregate
+	AllowedStatusNames = sets.NewString("alpha", "beta", "deprecated")
+
+	// regex for a valid k8s version
+	minVersionRegex = regexp.MustCompile(`^v[1-9][0-9]*\.(0|[1-9][0-9]*)$`)
+)
+
+// APILifecyleTag implements APIRule interface.
+type APILifecyleTag struct {
+	// FeatureGateFileNames is the list of files defining feature gates.
+	FeatureGateMap map[string]FeatureGateInfo
+}
+
+// FeatureGateInfo stores info about a particular feature gate.
+type FeatureGateInfo struct {
+	// status is the highest status for the feature gate.
+	status string
+	// minVersion is the minimum k8s version for the status.
+	minVersion string
+}
+
+// Name returns the name of APIRule
+func (l *APILifecyleTag) Name() string {
+	return "incorrect_lifecycle_tag"
+}
+
+// Validate evaluates API rule on type t and returns a list of field names in
+// the type that violate the rule. Empty field name [""] implies the entire
+// type violates the rule.
+func (l *APILifecyleTag) Validate(t *types.Type) ([]string, error) {
+	fields := make([]string, 0)
+
+	switch t.Kind {
+	case types.Struct:
+		for _, m := range t.Members {
+			commentTag := types.ExtractCommentTags("+", m.CommentLines)[TagLifecycle]
+			if commentTag != nil {
+				// only consider the first occurance of tagLifecycleComponent
+				// commentValues is of the form ["kubernetes", "minVersion=1.22", "status=alpha", "featureGate=Foo"]
+				commentValues := strings.Split(commentTag[0], ",")
+
+				// the lifecycle tag must have complete info
+				if len(commentValues) != 4 {
+					fields = append(fields, m.Name)
```

Comment:
I'm concerned how an API author would tell why a lifecycle tag is invalid, given different ways a tag may fail. Ideally a API rule name should be self explanatory. I wonder if `klog.Warning` would be too noisy. We also need to document what this rule checks in https://github.com/kubernetes/kubernetes/blob/master/api/api-rules/README.md

Reply:
> I'm concerned how an API author would tell why a lifecycle tag is invalid, given different ways a tag may fail. Ideally a API rule name should be self explanatory.

I agree. I was initially not sure if this was the right approach as well but splitting the linter for each "key" in the tag didn't feel clean either. :grimacing: 

> We also need to document what this rule checks in https://github.com/kubernetes/kubernetes/blob/master/api/api-rules/README.md

Will do that in the PR that will update kube-openapi. :+1: 

>  I wonder if klog.Warning would be too noisy. 

I have updated to use `klog.V(2)`. That's the log level we use for other errors in kube-openapi as well.

re: noisy - it _is_ noisy right now because most of the logs are from gengo. I don't think the gengo logs need to displayed at v=2 though, so I've created https://github.com/kubernetes/gengo/pull/207.

Once the gengo logs have been updated, `make generated_files UPDATE_API_KNOWN_VIOLATIONS=true KUBE_VERBOSE=2` (the command that we suggest to run in the [api-rules README](https://github.com/kubernetes/kubernetes/blob/master/api/api-rules/README.md)) still returns noisy logs because there are many v=2 logs in other generators.

I was wondering if this would be an acceptable solution. Wdyt?
- Increase log level for gengo wherever appropriate in https://github.com/kubernetes/gengo/pull/207
- Use v=2 for linter errors in kube-openapi
- Update the command in the [api-rules README](https://github.com/kubernetes/kubernetes/blob/master/api/api-rules/README.md) to just call openapi-gen i.e. `make gen_openapi UPDATE_API_KNOWN_VIOLATIONS=true`. If someone needs to display logs, they can add  `KUBE_VERBOSE=2`

With these approach, logs look something like this

<details><summary>Example logs</summary>
<p>

```console
make gen_openapi UPDATE_API_KNOWN_VIOLATIONS=true KUBE_VERBOSE=2
Go version: go version go1.16.4 linux/amd64
+++ [0629 20:08:44] Building go targets for linux/amd64:
    ./vendor/k8s.io/kube-openapi/cmd/openapi-gen
Env for linux/amd64: GOOS=linux GOARCH=amd64 GOROOT=/usr/local/go CGO_ENABLED= CC=
Coverage is disabled.
+++ [0629 20:09:06] Placing binaries
Generating openapi code for KUBE
I0629 20:09:09.579216 2563452 openapi.go:471] [k8s.io/api/apps/v1.DeploymentSpec] Strategy k8s.io/api/apps/v1.DeploymentStrategy: tag patchStrategy on type Struct; only allowed on type Slice
I0629 20:09:09.603351 2563452 openapi.go:471] [k8s.io/api/apps/v1beta1.DeploymentSpec] Strategy k8s.io/api/apps/v1beta1.DeploymentStrategy: tag patchStrategy on type Struct; only allowed on type Slice
I0629 20:09:09.641154 2563452 openapi.go:471] [k8s.io/api/apps/v1beta2.DeploymentSpec] Strategy k8s.io/api/apps/v1beta2.DeploymentStrategy: tag patchStrategy on type Struct; only allowed on type Slice
I0629 20:09:10.228820 2563452 openapi.go:471] [k8s.io/api/core/v1.PodSpec] Volumes []k8s.io/api/core/v1.Volume: [merge,retainKeys] not allowed for patchStrategy. Allowed values: [merge retainKeys]
I0629 20:09:10.346652 2563452 openapi.go:471] [k8s.io/api/extensions/v1beta1.DeploymentSpec] Strategy k8s.io/api/extensions/v1beta1.DeploymentStrategy: tag patchStrategy on type Struct; only allowed on type Slice
I0629 20:09:10.490956 2563452 openapi.go:471] [k8s.io/api/policy/v1.PodDisruptionBudgetSpec] Selector *k8s.io/apimachinery/pkg/apis/meta/v1.LabelSelector: [replace] not allowed for patchStrategy. Allowed values: [merge retainKeys]
I0629 20:09:10.490979 2563452 openapi.go:471] [k8s.io/api/policy/v1.PodDisruptionBudgetSpec] Selector *k8s.io/apimachinery/pkg/apis/meta/v1.LabelSelector: tag patchStrategy on type Pointer; only allowed on type Slice
I0629 20:09:10.497159 2563452 openapi.go:471] [k8s.io/api/policy/v1beta1.PodDisruptionBudgetSpec] Selector *k8s.io/apimachinery/pkg/apis/meta/v1.LabelSelector: [replace] not allowed for patchStrategy. Allowed values: [merge retainKeys]
I0629 20:09:10.497178 2563452 openapi.go:471] [k8s.io/api/policy/v1beta1.PodDisruptionBudgetSpec] Selector *k8s.io/apimachinery/pkg/apis/meta/v1.LabelSelector: tag patchStrategy on type Pointer; only allowed on type Slice
I0629 20:09:10.659323 2563452 openapi.go:471] [k8s.io/apimachinery/pkg/apis/meta/v1.LabelSelectorRequirement] Key string: tag patchMergeKey on type Builtin; only allowed on type Slice
I0629 20:09:10.659343 2563452 openapi.go:471] [k8s.io/apimachinery/pkg/apis/meta/v1.LabelSelectorRequirement] Key string: tag patchStrategy on type Builtin; only allowed on type Slice
I0629 20:09:10.869985 2563452 lifecycle_tag.go:103] [k8s.io/api/apps/v1.StatefulSetSpec] MinReadySeconds: "lifecycle:component" tag must have all keys in [component featureGate minVersion status], found: [kubernetes minVersion=v1.21 featureGate=StatefulSetMinReadySeconds]
Generating openapi code for AGGREGATOR
I0629 20:09:14.916339 2563561 openapi.go:471] [k8s.io/apimachinery/pkg/apis/meta/v1.LabelSelectorRequirement] Key string: tag patchMergeKey on type Builtin; only allowed on type Slice
I0629 20:09:14.916437 2563561 openapi.go:471] [k8s.io/apimachinery/pkg/apis/meta/v1.LabelSelectorRequirement] Key string: tag patchStrategy on type Builtin; only allowed on type Slice
Generating openapi code for APIEXTENSIONS
I0629 20:09:17.131654 2563646 openapi.go:471] [k8s.io/apimachinery/pkg/apis/meta/v1.LabelSelectorRequirement] Key string: tag patchMergeKey on type Builtin; only allowed on type Slice
I0629 20:09:17.131708 2563646 openapi.go:471] [k8s.io/apimachinery/pkg/apis/meta/v1.LabelSelectorRequirement] Key string: tag patchStrategy on type Builtin; only allowed on type Slice
Generating openapi code for CODEGEN
I0629 20:09:19.105826 2563731 openapi.go:471] [k8s.io/apimachinery/pkg/apis/meta/v1.LabelSelectorRequirement] Key string: tag patchMergeKey on type Builtin; only allowed on type Slice
I0629 20:09:19.105878 2563731 openapi.go:471] [k8s.io/apimachinery/pkg/apis/meta/v1.LabelSelectorRequirement] Key string: tag patchStrategy on type Builtin; only allowed on type Slice
Generating openapi code for SAMPLEAPISERVER
I0629 20:09:21.052467 2563816 openapi.go:471] [k8s.io/apimachinery/pkg/apis/meta/v1.LabelSelectorRequirement] Key string: tag patchMergeKey on type Builtin; only allowed on type Slice
I0629 20:09:21.052517 2563816 openapi.go:471] [k8s.io/apimachinery/pkg/apis/meta/v1.LabelSelectorRequirement] Key string: tag patchStrategy on type Builtin; only allowed on type Slice
```

</p>
</details>

Reply:
Thanks for the deep thoughts. The result looks really nice. 

I'm not sure if it's okay to change the log level in gengo. People may be depending on those behaviors. 

We can also add a flag in kube-openapi to log at warning / write to a file, so we don't need to worry about gengo logs.

That being said, it doesn't block this PR, assuming we will have good documentation in k/k.

Reply:
Updated to use `klog.Warningf` for now - this is not spammy at all because we don't use `Warningf` for any other warnings.

Having said that, I can update this to write to a file as a follow-up. :+1: